### ISSUE TITLE
Mongo 3.6 results in more transactions used and thus slower response than 3.2

### ISSUE THREAD

**Issue by MobileVet** [parse-community/parse-server#5735]
### Issue Description

We updated our mLab instance on May 22nd to utilize Mongo 3.6 as they removed support for 3.2 (our prior version).  We immediately saw a 20% increase in request queuing time, as reported by New Relic.  Looking at the mLab analytics, it appears related to the number of 'commands' issued and possibly the amount of cache evictions... but that is only a theory.

### Steps to reproduce

- Update from Mongo 3.2 to Mongo 3.6

### Expected Results

Negligible change in system performance

### Actual Outcome

- **20%** increase in server request queuing time (New Relic)
- **50%** increase in transaction checkpoint times (mLab)
- **40%** increase in memory use @ mLab
- **150%** increase in inbound network traffic to db (mLab)
- **Dramatic** decrease in 'selected but not evicted' pages in db (mLab)
  - from ~40 to 0
- **500%** increase in Mongo 'commands per second' (mLab)
  - mLab indicated this may be more in what they consider a 'command' in 3.6 vs 3.2

### Environment Setup

- **Server**
  - parse-server version:  2.8.2
  - Operating System:     Ubuntu 16.04. (Heroku 16)
  - Hardware:             Heroku Performance L - 14GB Ram, 100% CPU share
  - Remote server:  Heroku

- **Database**
  - MongoDB version: 3.6.12
  - Storage engine:  wiredTiger
  - Hardware:        M2 High storage, 150gb SSD
  - Remote server:  mLab

- **Stack**
  - Node: 8.14.0
  - Express: 4.16.3
  - Throng 4.0.0
  - Node clustering ENABLED w/ 20 worker processes (21 total processes)

- **Traffic**
  - Min load: 20 requests per second
  - Max load: 140 requests per second
  - Mean load: 100 requests per second

### Logs/Trace

 - See included graphs
<img width="697" alt="Web Transaction Time" src="https://user-images.githubusercontent.com/1977118/60187878-5ed3c380-97fc-11e9-9c07-ef4f61407efd.png">
<img width="402" alt="mLab Operations" src="https://user-images.githubusercontent.com/1977118/60187910-6eeba300-97fc-11e9-8225-28027c12b795.png">
<img width="399" alt="mLab Transaction Time" src="https://user-images.githubusercontent.com/1977118/60187950-80cd4600-97fc-11e9-827a-0ca0e09ed6bc.png">
<img width="399" alt="mLab Cache Evictions" src="https://user-images.githubusercontent.com/1977118/60188071-b1ad7b00-97fc-11e9-9d00-888f222928fe.png">
<img width="403" alt="mLab Network Throughput" src="https://user-images.githubusercontent.com/1977118/60188095-bd00a680-97fc-11e9-8602-6b073c6a70be.png">

<img width="398" alt="mLab Memory Use" src="https://user-images.githubusercontent.com/1977118/60188135-cc7fef80-97fc-11e9-95f6-f699c1322c8e.png">

**Comment by dplewis** [comment]
Can you try updating to the latest version of Parse Server and post results? A lot of cache improvements have been made.

**Comment by MobileVet** [comment]
Yea... I would love to move us up to the current server version but that would require a ton of work due to the breaking changes around **Parse.Cloud** function and **Parse.Promise**.  We plan to clean that up and make the transition... but it isn't something I can do quickly.

I just rolled us back to Mongo 3.4.  It is too early to see a difference in our Server response time as reported by New Relic, but the Mongo telemetry is tracking back to prior levels... memory, cache page evictions, operations etc.  Guessing the response times will do the same.

**Comment by MobileVet** [comment]
mLab has indicated that the change in # of 'commands' is based on how they are reported in 3.6 vs 3.4.

What is really interesting though is the large increase in network throughput INTO mLab.  Mongo 3.6 ends up getting 3x the amount of traffic INTO mLab than prior version.
<img width="515" alt="mLab Network Traffic IN" src="https://user-images.githubusercontent.com/1977118/60353702-92e3eb80-9998-11e9-9488-fe50d9d9506d.png">

Does it make sense that the Mongo adapter is sending more commands to support the same requests in 3.6 vs 3.4?  Would this result in more time for the total system endpoint?

**Comment by MobileVet** [comment]
One other thing we noticed, the **number** of Node garbage collections is the same but the aggregate time **doing** garbage collection is 30% **larger** when using Mongo 3.6 versus 3.4 or 3.2

**Comment by davimacedo** [comment]
The command sent by the Mongo Adapter should be the same for the two versions of Mongo. Maybe some option used by Mongo Adapter has changed from one version to another? I am not sure. Is there anyway for you to know which queries are generating more output traffic?

**Comment by MobileVet** [comment]
@davimacedo Hmm, that is a really good question.  I am not sure how to do that without spending a ton of time. Let me think about that....

In the mean time, I think you are likely onto something with the default options being changed across versions.  I have been chatting with the **mLab** team and they have shared several adjustments.

> Two big changes introduced with MongoDB 3.6 are [Change Streams](https://docs.mongodb.com/manual/changeStreams/) and [Server Sessions](https://docs.mongodb.com/manual/reference/server-sessions/)

I do see that the Parse Server has different versions of the Node adapter from 3.x to 2.8.2 (what we are using).  Those changes can be seen [here](https://github.com/mongodb/node-mongodb-native/blob/master/HISTORY.md)

- Is there any documentation regarding the Mongo Adapter options and their use in Parse beyond core [**mongodb** package API](http://mongodb.github.io/node-mongodb-native/2.1/reference/connecting/connection-settings/)?
- Is there a set of recommended options (for instance, 30s timeout seems awfully long)?

**Comment by davimacedo** [comment]
The version of the MongoDB Node.js driver that Parse Server uses is fixed in each Parse Server release. I've just checked [here](https://github.com/parse-community/parse-server/blob/0c2f1d583b7eadac08fcb0f9f5fdd9eeb1bedb24/package.json#L35) and Parse Server 2.8.2 uses MongoDB driver 3.0.7. The latest version of the driver is 3.2 so you may have been suffering from some incompatibility of running MongoDB 3.6 while using MongoDB Node.js 3.0.7 driver. I'd recommend you to either:
1) Upgrade Parse Server to the latest version - like @dplewis suggested before
2) Try to fork the Parse Server 2.8.2, upgrade only the MongoDB driver, run the unit tests, and then run your app from your fork

**Comment by MobileVet** [comment]
@davimacedo Thanks for the follow up.

I noticed the version issue with MongoDB as well, I think you might be on to something there.  I asked the folks over at mLab, will be interesting to see what they think.

We will be updating to the latest Parse version, but probably not for another couple months.  In the mean time I will give your other suggestion a show (forking 2.8.2 and updating the MongoDB driver).

We have rolled back to 3.4 and our top level performance statistics as reported by New Relic have returned to pre 3.6 levels.
